score.go

```diff
@@ -0,0 +1,952 @@
+	if d.head == nil {
+		return
+	}
+
+	now := time.Now()
+	for d.head != nil && now.After(d.head.expire) {
+		delete(d.records, d.head.id)
+		d.head = d.head.next
+	}
+
+	if d.head == nil {
+		d.tail = nil
+	}
+}
+
+// utilities
+func (pstats *peerStats) getTopicStats(topic string, params *PeerScoreParams) (*topicStats, bool) {
+	tstats, ok := pstats.topics[topic]
+	if ok {
+		return tstats, true
+	}
+
+	_, scoredTopic := params.Topics[topic]
+	if !scoredTopic {
+		return nil, false
+	}
+
+	tstats = &topicStats{}
+	pstats.topics[topic] = tstats
+
+	return tstats, true
+}
+
+func (ps *peerScore) markInvalidMessageDelivery(p peer.ID, msg *Message) {
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return
+	}
+
+	for _, topic := range msg.GetTopicIDs() {
+		tstats, ok := pstats.getTopicStats(topic, ps.params)
+		if !ok {
+			continue
+		}
+
+		tstats.invalidMessageDeliveries += 1
+	}
+}
+
+func (ps *peerScore) markFirstMessageDelivery(p peer.ID, msg *Message) {
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return
+	}
+
+	for _, topic := range msg.GetTopicIDs() {
+		tstats, ok := pstats.getTopicStats(topic, ps.params)
+		if !ok {
+			continue
+		}
+
+		cap := ps.params.Topics[topic].FirstMessageDeliveriesCap
+		tstats.firstMessageDeliveries += 1
+		if tstats.firstMessageDeliveries > cap {
+			tstats.firstMessageDeliveries = cap
+		}
+
+		if !tstats.inMesh {
+			continue
+		}
+
+		cap = ps.params.Topics[topic].MeshMessageDeliveriesCap
+		tstats.meshMessageDeliveries += 1
+		if tstats.meshMessageDeliveries > cap {
+			tstats.meshMessageDeliveries = cap
+		}
+	}
+}
+
+func (ps *peerScore) markDuplicateMessageDelivery(p peer.ID, msg *Message, validated time.Time) {
+	var now time.Time
+
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return
+	}
+
+	if !validated.IsZero() {
+		now = time.Now()
+	}
+
+	for _, topic := range msg.GetTopicIDs() {
+		tstats, ok := pstats.getTopicStats(topic, ps.params)
+		if !ok {
+			continue
+		}
+
+		if !tstats.inMesh {
+			continue
+		}
+
+		// check against the mesh delivery window -- if the validated time is passed as 0, then
+		// the message was received before we finished validation and thus falls within the mesh
+		// delivery window.
+		if !validated.IsZero() && now.After(validated.Add(ps.params.Topics[topic].MeshMessageDeliveriesWindow)) {
+			continue
+		}
+
+		cap := ps.params.Topics[topic].MeshMessageDeliveriesCap
+		tstats.meshMessageDeliveries += 1
+		if tstats.meshMessageDeliveries > cap {
+			tstats.meshMessageDeliveries = cap
+		}
+	}
+}
+
+// gets the current IPs for a peer
+func (ps *peerScore) getIPs(p peer.ID) []string {
+	// in unit tests this can be nil
+	if ps.host == nil {
+		return nil
+	}
+
+	conns := ps.host.Network().ConnsToPeer(p)
+	res := make([]string, 0, len(conns))
+	for _, c := range conns {
+		remote := c.RemoteMultiaddr()
+
+		ip4, err := remote.ValueForProtocol(ma.P_IP4)
+		if err == nil {
+			res = append(res, ip4)
+			continue
+		}
+
+		ip6, err := remote.ValueForProtocol(ma.P_IP6)
```

Comment:
Note: this function is dangerous as it'll pick out the first IP address, regardless of what comes first. I can't think of a case where it'll fail at the moment, but it's best to avoid it and call something like `manet.ToIP`.

Reply:
ok, can do.

Reply:
done.